designdocs/MESSAGE_STORAGE.md

```diff
@@ -47,9 +47,9 @@ using [StorageDoubleMap](https://docs.substrate.io/rustdocs/latest/frame_support
 Following is a proposed data structure for storing a Message on chain.
 ```rust
 pub struct Message<AccountId> {
-    pub data: Vec<u8>,		    //  Serialized data in a user-defined schemas format
-    pub signer: AccountId,	    //  Signature of the signer
-    pub msa_id: u64,                //  Message source account id (the original sender)
+    pub payload: Vec<u8>,		    //  Serialized data in a user-defined schemas format
+    pub provider_key: AccountId,	    //  Signature of the signer
```

Comment:
I think you mentioned there are two types of provider:
1. registered provider
2. Provider 

A provider is some one who has capacity but cannot delegate?
A registered provider is someone who can delegate and is verified?

Reply:
True. This can be either. #129 talks about registering providers

Reply:
In this case this would be a registered provider since we have a delegation validation.

Reply:
It *might* be a registered provider. Since `on_behalf_of` is Optional. If `on_behalf_of` is `Some` then it must be a registered provider. Otherwise `msa_id` is the id from the `provider_key`.

It is a bit confusing as we have that msa_id as optional. I'm open for ideas on how to reduce that confusion.

Here's how it works in my head:
- Every `Message` must have a `provider` (stored via the `provider_key`). The provider of the capacity that sent the message.
- Every message with an on-chain payload will have a source (the `msa_id`) which might be the msa_id of the provider, but could be a validated point in time delegation
- Every message with off-chain payload will ONLY have a provider. (The source, if any, must be in the off-chain payload and validated at read time)

Another way to put it

| -- | Provider | Sender |
| --- | --- | --- |
| Referenced via | key | msa_id |
| On-chain Msg Self Provider | Required | Same as Provider |
| On-chain Msg Delegated | Required | Required, Validated Delegation |
| Off-chain Msg | Required | Optional & Off-chain |

Reply:
If on_behalf_of is some we know its a delegation, now about msa_id being optional, at the code level per se we can define bunch of macros 
However we can also enforce to always include on behalf of in cases its same as origin we know its provider else registered provider